### PR TITLE
Confirm order

### PR DESCRIPTION
**Pull request by Ru-T**
## What Changed?
* Added a preview step before purchase of a book order
* Added partial with preview step info
* Passed feature specs for confirmation step of purchase tests

## Why? 
* So a user can confirm their order before being charged

### CODE REVIEW

**Comment by jahammo2** [comment]
So I pulled this branch ran the cucumber tests. It seems you never defined what `last4` is supposed to be for an order. Is this happening for you?
```
Feature: Purchasing a Book

  @javascript
  Scenario: Happy Path                                                                                  # features/purchase.feature:3
    Given I am logged into the site                                                                     # features/step_definitions/administration_steps.rb:36
    And there are some books in the database                                                            # features/step_definitions/purchase_steps.rb:5
    When I visit the public book index                                                                  # features/step_definitions/administration_steps.rb:86
    And I click on a book                                                                               # features/step_definitions/purchase_steps.rb:11
    And I click "Add to Cart"                                                                           # features/step_definitions/purchase_steps.rb:15
    When I visit my cart                                                                                # features/step_definitions/purchase_steps.rb:27
    Then I see the book in my cart                                                                      # features/step_definitions/purchase_steps.rb:42
    When I click "Purchase"                                                                             # features/step_definitions/purchase_steps.rb:15
    Then I am asked for my shipping address                                                             # features/step_definitions/purchase_steps.rb:47
    When I enter a valid shipping address                                                               # features/step_definitions/purchase_steps.rb:51
    Then I am asked for my billing address                                                              # features/step_definitions/purchase_steps.rb:58
    When I enter a valid billing address                                                                # features/step_definitions/purchase_steps.rb:62
    Then I am asked for my credit card                                                                  # features/step_definitions/purchase_steps.rb:69
    When I enter a valid credit card                                                                    # features/step_definitions/purchase_steps.rb:73
    And I click "Submit Order"                                                                          # features/step_definitions/purchase_steps.rb:15
    Then I am asked to review the order total                                                           # features/step_definitions/purchase_steps.rb:84
      undefined method `last4' for nil:NilClass (NoMethodError)
      ./app/models/order.rb:29:in `save_card'
      ./app/controllers/orders_controller.rb:18:in `create'
      features/purchase.feature:19:in `Then I am asked to review the order total'
```

**Comment by jahammo2** [comment]
Oh, it's a Stripe thing. Still, are you getting this error?

**Comment by Ru-T** [comment]
@jahammo2 Nope - all tests passing!

**Comment by Ru-T** [comment]
screen shot 2016-02-09 at 12 21 24 pm

**Comment by jahammo2** [comment]
Interesting, well no biggie

File: app/controllers/orders_controller.rb [hunk 1]
```diff
@@ -16,12 +16,16 @@ def create
...
+      render partial: "preview"
     else
-      redirect to cart_path(current_user), notice: "Your order could not be processed."
+      if @order.save_with_payment
+        @order.purchase_line_items
+        OrderMailer.receipt_email(@order.user).deliver_now
+        redirect_to order_path(@order), notice: "Your order has been completed"
+      else
+        redirect to cart_path(current_user), notice: "Your order could not be processed."
+      end
```
Comment (jahammo2) [inline app/controllers/orders_controller.rb]: So this method is starting to have a lot of responsibility and contains a lot of conditionals. It's definitely time for a refactor. If I may suggest:

```
class OrdersController < ApplicationController
class OrdersController < ApplicationController
  before_action :authenticate_user!
  before_action :set_order, only: [:show]

  def new
    @order = Order.new(user: current_user)
    @order.user.credit_card ||= @order.user.build_credit_card
  end

  def create
    @order = Order.new(
      order_params.merge(
        user: current_user,
        total: current_user.cart.total_cart_price,
        stripe_token: params[:stripeToken]
      )
    )
    prepare_order(@order)
  end

  private

  def set_order
    @order = Order.find(params[:id])
  end

  def prepare_order(order)
    save_card_to_order(order)
    render_order(order)
  end

  def save_card_to_order(order)
    order.save_card if current_user.credit_card.card_token.nil?
  end

  def render_order(order)
    if params[:preview]
      render partial: "preview"
    else
      save_order(order)
    end
  end

  def save_order(order)
    if order.save_with_payment
      handle_saved_order(order)
    else
      redirect to cart_path(current_user), notice: "Your order could not be processed."
    end
  end

  def handle_saved_order(order)
    order.purchase_line_items
    OrderMailer.receipt_email(order.user).deliver_now
    redirect_to order_path(order), notice: "Your order has been completed"
  end

  # strong params
  def order_params
    params.require(:order).permit(
      :user_id,
      :total,
      :stripe_token,
      :shipping_address1,
      :shipping_address2,
      :shipping_city,
      :shipping_state,
      :shipping_zip,
      credit_card_attributes: [
        :id,
        :billing_address1,
        :billing_address2,
        :billing_city,
        :billing_state,
        :billing_zip,
        :last_four_digits,
        :card_token
      ]
    )
  end
end
```

Everything still passes according to your Cucumber and RSpec tests. But I didn't do anything too special. 
1. I simply passed the instance variable as an argument so you don't have to keep typing `order` out as an instance variable.
2. Most importantly, I tried to decouple the responsibilities of your method in order to suit Sandi's :pray: Single Responsibility methodology.

This is definitely not the cleanest way to do this. You really should be handling this via services (don't worry about that for this project). I definitely believe this needs a refactor but don't feel like you have to take my suggested code.

If you have any questions about this, please ask!

**Comment by jahammo2** [comment]
Definitely refactor that method. Afterwards, :shipit:. Everything else looks gucci.

**Pull request merged**